<Menu /> component

lluia opened the pull request:
![captura de pantalla 2018-11-20 a las 22 17 28](https://user-images.githubusercontent.com/5938217/48803393-160dd680-ed12-11e8-931d-868595188694.png)

Beginning of September, I volunteered to [work on a `<Menu />` component](https://github.com/reakit/reakit/issues/90#issuecomment-419972448). Unfortunately, travelling and recovering from it moved that contribution to the back of my mind.

I noticed that [the issue is still open](https://github.com/reakit/reakit/issues/90) so I assumed it's still relevant, so nevertheless I had some fun 🍿

A bit late, but better late than never! 🦄 💁🏻‍♂️

## What kind of change does this PR introduce?

Adds a new `<Menu />` component to **Reakit** with the following API:
```jsx
<Menu>
  <Menu.Item>Item 1</Menu.Item>
  <Menu.Item>Item 2</Menu.Item>
  <Menu.Divider />
  <Menu.Item>Item 3</Menu.Item>
</Menu>
```
The layout can be controlled via `horizontal` and `horizontalAt` props as per [this suggestion](https://github.com/reakit/reakit/issues/90#issuecomment-420058411).

## Does this PR introduce a breaking change?

I bet this fits into a _minor version_ bump 🏷

## Notes for the reviewer

#### 1
I'm not familiar with **Typescript**, I did my best, having extra &nbsp; 👀 &nbsp;on the type declarations would be 👌🏻

#### 2
I abstracted [this behavior](https://github.com/reakit/reakit/blob/master/packages/reakit/src/Group/Group.ts#L16-L28) from `<Group />` into a common utility to not repeat the same in `<Menu />` and stay consistent. 

#### 3
To make `<MenuDivider />` implicitly aware of whether the menu is configured as `horizontal` or not ( as it needs to render in one way or another accordingly ) I used **React's Context**. 

#### 4
Should `<Menu />` accept any type of `children` or only `<MenuItem />` or `<MenuDivider />` ?

PlayMa256 commented on packages/reakit/src/Menu/Menu.tsx:
```diff
@@ -0,0 +1,40 @@
+import React from "react";
```
```suggestion
import * as React from "react";
```

lluia replied:
Happy to amend here and below 🙌🏻

Out of curiosity, why there's a need to access all the exports of **React** behind a namespace? is it a convention or something related to **Typescript**? 

Can it be something we could set the linter to catch? ( happy to add that config in this PR if _positivie_ )

PlayMa256 replied:
That is due to not having `allowSyntheticDefaultImports` on our `tsconfig.json` file. I don't know the implications of having or not this, but i know for sure that the lack of this implies importing things with `*`.

diegohaz replied:
It would throw errors on users' projects (which is weird, but ok). See https://github.com/reakit/reakit/issues/198#issuecomment-414009701

PlayMa256 commented on packages/reakit/src/Menu/MenuDivider.tsx:
```diff
@@ -0,0 +1,42 @@
+import React from "react";
```
```suggestion
import * as React from "react";
```

PlayMa256 commented:
Those were my 2 cents. Behavior is ok to me! Can you address the type errors from ci? If you have any problem just let me know.

lluia commented on packages/reakit/src/Menu/Menu.tsx:
```diff
@@ -0,0 +1,40 @@
+  };
+
+  return (
+    <HorizontalMenuContext.Provider value={layout}>
```
We're already spreading the props to `<List />` which is the lowest level child `<Menu />` knows about:
```jsx
function Menu(props: MenuProps) {
   // ...
   <List {...props} />
   // ...
}
```
To make it work the way you propose we would need to modify `<List />` so that it also spreads its props down to its `children`.

However that would imply tying the behaviour of one component to the other and end with tight coupling between the two 🧐 👎🏻

PlayMa256 replied:
Yeah yeah... big nono 💥

diegohaz replied:
Let's try to render only one component. Context isn't needed. Just add [aria-orientation](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-orientation_attribute) and style its children based on it. For example:
```
const MenuDivider = styled(Box)`
  [aria-orientation="horizontal"] > & {
    // styles for vertical
  }
`;
```

lluia commented on packages/reakit/src/Menu/MenuItem.ts:
```diff
@@ -0,0 +1,20 @@
+  -webkit-user-select: none; /* Chrome all / Safari all */
+  -moz-user-select: none; /* Firefox all */
+  -ms-user-select: none; /* IE 10+ */
+  user-select: none; /* Likely future */
```
Good point, styled components prefix automatically so no need to have this here 🎈

diegohaz commented on packages/reakit/src/Menu/MenuDivider.tsx:
```diff
@@ -0,0 +1,42 @@
+  )};
+
+  background-color: ${p("background", 1)};
+  ${theme("MenuItem")};
```
```suggestion
  ${theme("MenuDivider")};
```

PlayMa256 replied:
good catch

lluia replied:
ha ha pushing code at night 🙈

diegohaz commented on packages/reakit/src/Group/Group.md:
```diff
@@ -6,7 +6,7 @@ import { Button, Group } from "reakit";
   <Button>Ok</Button>
   <Button>Cancel</Button>
-</Group>
+</Group>;
```
Prettier, I know 😄
```suggestion
</Group>
```

diegohaz commented on packages/reakit/src/Group/Group.md:
```diff
@@ -17,5 +17,5 @@ import { Button, Group } from "reakit";
   <Button maxWidth="20vmin">Up</Button>
   <Button maxWidth="20vmin">Down</Button>
-</Group>
+</Group>;
```
```suggestion
</Group>
```

diegohaz replied:
Oh yes! Didn't know about that option.

diegohaz commented on packages/reakit/src/Menu/Menu.md:
````diff
@@ -0,0 +1,12 @@
+A menu that does nothing for now...
+
+```jsx
+<Menu>
````
```suggestion
import { Menu } from "reakit";

<Menu>
```

lluia replied:
Oh interesting, @diegohaz you know why I didn't see any error when rendering this locally due to the missing `import`?

diegohaz replied:
Technically, those imports aren't needed (styleguidist already creates global variables for them). But we started to adopt it as a convention in #276 so as to make it clearer for people reading the docs.

diegohaz replied:
Maybe we should find a way to disable that auto import on styleguidist so we can catch missing imports more easily. I'm not sure if it's possible though.

diegohaz commented on packages/reakit/src/Menu/Menu.md:
```diff
@@ -0,0 +1,12 @@
+<Menu>
+  <Menu.Item onClick={() => alert("Hello!")}>Item 1</Menu.Item>
+  <Menu.Item>Item 2</Menu.Item>
+  <Menu.Divider style={{ backgroundColor: "#eee" }} />
```
```suggestion
  <Menu.Divider backgroundColor="#eee" />
```

lluia commented:
Thanks a lot for the feedback @diegohaz @PlayMa256, I addressed your suggestions and everything looks much lean 💖 🎊 , &nbsp;👀&nbsp; my [last commit](https://github.com/reakit/reakit/pull/290/commits/8a0e3ff4b316df8a74211b9cb8ae44ca4f1009e0) ( _force pushed my own branch for a clearer history_ ).

I improved the README for the new component and fixed the linting errors 🤞🏻 

Are we ok with the default styles set on the new components? ( _I see most of the other components have no styles and leave the theme to actually define that_ )

lluia commented:
@diegohaz any feedback on the updates ☝🏼 ? no rush from my side, I'm happy to wait or make more amendments 👍🏻

diegohaz commented on packages/reakit/src/Menu/Menu.tsx:
```diff
@@ -0,0 +1,18 @@
+
+  ${theme("Menu")};
+`;
+
```
I think it's better to accept a `vertical` prop and set `aria-orientation` accordingly. Thus we keep it consistent with other components like [`Toolbar`](https://github.com/reakit/reakit/blob/master/packages/reakit/src/Toolbar/Toolbar.tsx).

Also, it needs a `role` in `defaultProps`:
```suggestion

Menu.defaultProps = {
  role: "menu"
};
```

lluia replied:
Sounds cool, I guess you meant to name the prop `horizontal` rather than `vertical` ( _right now we're rendering the menu vertically and through a prop we make it horizontal_ ) unless you prefer to do the inverse ?

diegohaz replied:
You're right. `horizontal` is better.

diegohaz commented on packages/reakit/src/Menu/MenuDivider.tsx:
```diff
@@ -0,0 +1,21 @@
+  ${theme("MenuDidiver")};
+`;
+
+export default as("div")(MenuDivider);
```
`MenuDivider` should be an `li` and have a `role="separator"` prop (so screen readers will announce it correctly):
```suggestion
MenuDivider.defaultProps = {
  role: "separator"
};

export default as("li")(MenuDivider);
```

diegohaz commented on packages/reakit/src/Menu/MenuItem.ts:
```diff
@@ -0,0 +1,20 @@
+import as from "../as";
+import Box, { BoxProps } from "../Box";
+
+type Props = {
```
```suggestion
export interface MenuItemProps extends BoxProps {
```
See https://github.com/reakit/reakit/blob/master/CONTRIBUTING.md#components

diegohaz commented on packages/reakit/src/Menu/MenuItem.ts:
```diff
@@ -0,0 +1,20 @@
+  disabled?: boolean;
+};
+
+const MenuItem = styled(Box)<Props & BoxProps>`
```
```suggestion
const MenuItem = styled(Box)<MenuItemProps>`
```

diegohaz commented on packages/reakit/src/Menu/MenuItem.ts:
```diff
@@ -0,0 +1,20 @@
+  ${theme("MenuItem")};
+`;
+
+export default as("li")(MenuItem);
```
```suggestion
MenuItem.defaultProps = {
  role: "menuitem"
};

export default as("li")(MenuItem);
```

lluia commented:
@diegohaz addressed your suggestions, good catch adding `role` to each component 🙌🏻
Happy to refine more if needed 👍🏻

lluia commented:
@diegohaz _I think_ this one _looks ready_, now that I know better how the project is structured hopefully next PR ends up more lean 💅🏻 

In case we want to refine it more I may close this one and open a new one linking to it, as it's becoming a bit hard to manage feedback wise 👍🏻

diegohaz commented:
This looks good to me @lluia. Thank you very much for the awesome work. I'm gonna wait a little bit to merge this because we need to release a few things first.

jxom commented:
Should we also consider [keyboard interaction](https://www.w3.org/TR/wai-aria-practices-1.1/#menu) on the menu in order for it to be accessible? Unfortunately, by default, I don't think browsers come with this keyboard interaction out of the box. 😞

lluia commented:
@jxom good tip, I'll try to add this during the weekend as seems a good idea 👌

lluia commented:
@jxom @diegohaz I'm not too familiar with keyboard interactions, I noticed a lot of it has to do where how it works with _sub-menus_ (a menu within a menu) is this something we intend to support on this component? (as currently it's built to be just one level?)

In case we want to support submenus we need to discuss the API for it?

lluia commented:
Closing this one as I believe is not relevant anymore ( see my comment #345 )
